File: test_plugin/echo.c
#include "../rx/include/rx_api.h"
#include <stdio.h>

static void echo_handler(void* ctx, rx_value_t msg, rx_scheduler_t* sched) {
    (void)ctx; (void)sched;
    fprintf(stderr, "Echo port received a message! tag=%llu\n",
            (unsigned long long)(msg.bits & RX_TAG_MASK));
}

void rx_load(rx_scheduler_t* sched) {
    uint32_t pid = rx_spawn_port(sched, NULL, echo_handler, NULL);
    if (pid != 0) rx_register_port(sched, "echo", pid);
}

File: test_plugin/echo.zig
const std = @import("std");
const c = @cImport(@cInclude("rx_api.h"));

fn echoHandler(ctx: ?*anyopaque, msg: c.rx_value_t, sched: ?*anyopaque) callconv(.c) void {
    _ = ctx;
    _ = sched;
    std.debug.print("Echo (Zig) port received a message! bits=0x{x}\n", .{msg.bits});
}

export fn rx_load(sched: ?*anyopaque) void {
    const pid = c.rx_spawn_port(sched, null, echoHandler, null);
    if (pid != 0) c.rx_register_port(sched, "echo", pid);
}
